File: src/net/dinkla/raytracer/math/AffineTransformation.java
package net.dinkla.raytracer.math;

import static net.dinkla.raytracer.math.MathUtils.PI_ON_180;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 28.04.2010
 * Time: 22:32:53
 * To change this template use File | Settings | File Templates.
 */
public class AffineTransformation {

    public Matrix forwardMatrix;
    public Matrix invMatrix;

    public AffineTransformation() {
        forwardMatrix = new Matrix();
        invMatrix = new Matrix();
    }
    
    public void translate(Vector3D v) {
        translate(v.getX(), v.getY(), v.getZ());
    }

    public void translate(final float x, final float y, final float z) {
        Matrix invTranslationMatrix = new Matrix();
        invTranslationMatrix.getM()[0][3] = - x;
        invTranslationMatrix.getM()[1][3] = - y;
        invTranslationMatrix.getM()[2][3] = - z;
        invMatrix = invMatrix.mult(invTranslationMatrix);

        Matrix translationMatrix = new Matrix();
        translationMatrix.getM()[0][3] = x;
        translationMatrix.getM()[1][3] = y;
        translationMatrix.getM()[2][3] = z;
        forwardMatrix = translationMatrix.mult(forwardMatrix);
    }

    public void scale(Vector3D v) {
        scale(v.getX(), v.getY(), v.getZ());
    }

    public void scale(final float x, final float y, final float z) {
        Matrix invScalingMatrix = new Matrix();
        invScalingMatrix.getM()[0][0] = 1.0f / x;
        invScalingMatrix.getM()[1][1] = 1.0f / y;
        invScalingMatrix.getM()[2][2] = 1.0f / z;
        invMatrix = invMatrix.mult(invScalingMatrix);

        Matrix scalingMatrix = new Matrix();
        scalingMatrix.getM()[0][0] = x;
        scalingMatrix.getM()[1][1] = y;
        scalingMatrix.getM()[2][2] = z;
        forwardMatrix = scalingMatrix.mult(forwardMatrix);        
    }

    public void rotateX(final float phi) {
        final float cosPhi = (float) Math.cos(phi * PI_ON_180);
        final float sinPhi = (float) Math.sin(phi * PI_ON_180);

        Matrix invRotationMatrix = new Matrix();
        invRotationMatrix.getM()[1][1] = cosPhi;
        invRotationMatrix.getM()[1][2] = sinPhi;
        invRotationMatrix.getM()[2][1] = - sinPhi;
        invRotationMatrix.getM()[2][2] = cosPhi;
        invMatrix = invMatrix.mult(invRotationMatrix);

        Matrix rotationMatrix = new Matrix();
        rotationMatrix.getM()[1][1] = cosPhi;
        rotationMatrix.getM()[1][2] = - sinPhi;
        rotationMatrix.getM()[2][1] = sinPhi;
        rotationMatrix.getM()[2][2] = cosPhi;
        forwardMatrix = rotationMatrix.mult(forwardMatrix);
    }

    public void rotateY(final float phi) {
        final float cosPhi = (float) Math.cos(phi * PI_ON_180);
        final float sinPhi = (float) Math.sin(phi * PI_ON_180);

        Matrix invRotationMatrix = new Matrix();
        invRotationMatrix.getM()[2][2] = cosPhi;
        invRotationMatrix.getM()[0][2] = - sinPhi;
        invRotationMatrix.getM()[2][0] = sinPhi;
        invRotationMatrix.getM()[0][0] = cosPhi;
        invMatrix = invMatrix.mult(invRotationMatrix);

        Matrix rotationMatrix = new Matrix();
        rotationMatrix.getM()[2][2] = cosPhi;
        rotationMatrix.getM()[0][2] = sinPhi;
        rotationMatrix.getM()[2][0] = - sinPhi;
        rotationMatrix.getM()[0][0] = cosPhi;
        forwardMatrix = rotationMatrix.mult(forwardMatrix);
    }

    public void rotateZ(final float phi) {
        final float cosPhi = (float) Math.cos(phi * PI_ON_180);
        final float sinPhi = (float) Math.sin(phi * PI_ON_180);

        Matrix invRotationMatrix = new Matrix();
        invRotationMatrix.getM()[0][0] = cosPhi;
        invRotationMatrix.getM()[0][1] = sinPhi;
        invRotationMatrix.getM()[1][0] = - sinPhi;
        invRotationMatrix.getM()[1][1] = cosPhi;
        invMatrix = invMatrix.mult(invRotationMatrix);

        Matrix rotationMatrix = new Matrix();
        rotationMatrix.getM()[0][0] = cosPhi;
        rotationMatrix.getM()[0][1] = - sinPhi;
        rotationMatrix.getM()[1][0] = sinPhi;
        rotationMatrix.getM()[1][1] = cosPhi;
        forwardMatrix = rotationMatrix.mult(forwardMatrix);
    }

    public void shear(Matrix s) {
        Matrix invShearingMatrix = new Matrix();
        
        // discriminant
    	float d = 1.0f - s.getM()[1][0] * s.getM()[0][1]
                       - s.getM()[2][0] * s.getM()[0][2]
                       - s.getM()[2][1] * s.getM()[1][2]
					   + s.getM()[1][0] * s.getM()[2][1] * s.getM()[0][2]
                       + s.getM()[2][0] * s.getM()[0][1] * s.getM()[2][1];

	    // diagonals
        invShearingMatrix.getM()[0][0] = 1.0f - s.getM()[2][1] * s.getM()[1][2];
        invShearingMatrix.getM()[1][1] = 1.0f - s.getM()[2][0] * s.getM()[0][2];
        invShearingMatrix.getM()[2][2] = 1.0f - s.getM()[1][0] * s.getM()[0][1];
        invShearingMatrix.getM()[3][3] = d;

	    // first row
	    invShearingMatrix.getM()[0][1] = -s.getM()[1][0] + s.getM()[2][0] * s.getM()[1][2];
	    invShearingMatrix.getM()[0][2] = -s.getM()[2][0] + s.getM()[1][0] * s.getM()[2][1];

	    // second row
	    invShearingMatrix.getM()[1][0] = -s.getM()[0][1] + s.getM()[2][1] * s.getM()[0][2];
	    invShearingMatrix.getM()[1][2] = -s.getM()[2][1] + s.getM()[2][0] * s.getM()[0][1];

	    // third row
	    invShearingMatrix.getM()[2][0] = -s.getM()[0][2] + s.getM()[0][1] * s.getM()[1][2];
	    invShearingMatrix.getM()[2][1] = -s.getM()[1][2] + s.getM()[1][0] * s.getM()[0][2];

	    // divide by discriminant
	    invShearingMatrix = invShearingMatrix.div(d);

    	invMatrix = invMatrix.mult(invShearingMatrix);
	    forwardMatrix = s.mult(forwardMatrix); 
    }
    
}

File: src/main/kotlin/net/dinkla/raytracer/math/AffineTransformation.kt
package net.dinkla.raytracer.math

import net.dinkla.raytracer.math.MathUtils.PI_ON_180
import net.dinkla.raytracer.utilities.equals
import net.dinkla.raytracer.utilities.hash
import kotlin.math.cos
import kotlin.math.sin

class AffineTransformation : Transformation {

    override var forwardMatrix: Matrix = Matrix.identity()
    override var invMatrix: Matrix = Matrix.identity()

    override fun translate(v: Vector3D) {
        translate(v.x, v.y, v.z)
    }

    override fun translate(x: Double, y: Double, z: Double) {
        val invTranslationMatrix = Matrix.identity()
        invTranslationMatrix[0, 3] = -x
        invTranslationMatrix[1, 3] = -y
        invTranslationMatrix[2, 3] = -z
        invMatrix *= invTranslationMatrix

        val translationMatrix = Matrix.identity()
        translationMatrix[0, 3] = x
        translationMatrix[1, 3] = y
        translationMatrix[2, 3] = z
        forwardMatrix = translationMatrix * forwardMatrix
    }

    override fun scale(v: Vector3D) {
        scale(v.x, v.y, v.z)
    }

    override fun scale(x: Double, y: Double, z: Double) {
        val invScalingMatrix = Matrix.identity()
        invScalingMatrix[0, 0] = 1.0 / x
        invScalingMatrix[1, 1] = 1.0 / y
        invScalingMatrix[2, 2] = 1.0 / z
        invMatrix *= invScalingMatrix


        val scalingMatrix = Matrix.identity()
        scalingMatrix[0, 0] = x
        scalingMatrix[1, 1] = y
        scalingMatrix[2, 2] = z
        forwardMatrix = scalingMatrix * forwardMatrix
    }

    override fun rotateX(phi: Double) {
        val cosPhi = cos(phi * PI_ON_180)
        val sinPhi = sin(phi * PI_ON_180)

        val invRotationMatrix = Matrix.identity()
        invRotationMatrix[1, 1] = cosPhi
        invRotationMatrix[1, 2] = sinPhi
        invRotationMatrix[2, 1] = -sinPhi
        invRotationMatrix[2, 2] = cosPhi
        invMatrix *= invRotationMatrix

        val rotationMatrix = Matrix.identity()
        rotationMatrix[1, 1] = cosPhi
        rotationMatrix[1, 2] = -sinPhi
        rotationMatrix[2, 1] = sinPhi
        rotationMatrix[2, 2] = cosPhi
        forwardMatrix = rotationMatrix * forwardMatrix
    }

    override fun rotateY(phi: Double) {
        val cosPhi = cos(phi * PI_ON_180)
        val sinPhi = sin(phi * PI_ON_180)

        val invRotationMatrix = Matrix.identity()
        invRotationMatrix[2, 2] = cosPhi
        invRotationMatrix[0, 2] = -sinPhi
        invRotationMatrix[2, 0] = sinPhi
        invRotationMatrix[0, 0] = cosPhi
        invMatrix *= invRotationMatrix

        val rotationMatrix = Matrix.identity()
        rotationMatrix[2, 2] = cosPhi
        rotationMatrix[0, 2] = sinPhi
        rotationMatrix[2, 0] = -sinPhi
        rotationMatrix[0, 0] = cosPhi
        forwardMatrix = rotationMatrix * forwardMatrix
    }

    override fun rotateZ(phi: Double) {
        val cosPhi = cos(phi * PI_ON_180)
        val sinPhi = sin(phi * PI_ON_180)

        val invRotationMatrix = Matrix.identity()
        invRotationMatrix[0, 0] = cosPhi
        invRotationMatrix[0, 1] = sinPhi
        invRotationMatrix[1, 0] = -sinPhi
        invRotationMatrix[1, 1] = cosPhi
        invMatrix *= invRotationMatrix

        val rotationMatrix = Matrix.identity()
        rotationMatrix[0, 0] = cosPhi
        rotationMatrix[0, 1] = -sinPhi
        rotationMatrix[1, 0] = sinPhi
        rotationMatrix[1, 1] = cosPhi
        forwardMatrix = rotationMatrix * forwardMatrix
    }

    override fun shear(s: Matrix) {
        var invShearingMatrix = Matrix.identity()

        // discriminant
        val d = ((1.0 - s[1, 0] * s[0, 1]
                - s[2, 0] * s[0, 2]
                - s[2, 1] * s[1, 2])
                + s[1, 0] * s[2, 1] * s[0, 2]
                + s[2, 0] * s[0, 1] * s[2, 1])

        // diagonals
        invShearingMatrix[0, 0] = 1.0 - s[2, 1] * s[1, 2]
        invShearingMatrix[1, 1] = 1.0 - s[2, 0] * s[0, 2]
        invShearingMatrix[2, 2] = 1.0 - s[1, 0] * s[0, 1]
        invShearingMatrix[3, 3] = d

        // first row
        invShearingMatrix[0, 1] = -s[1, 0] + s[2, 0] * s[1, 2]
        invShearingMatrix[0, 2] = -s[2, 0] + s[1, 0] * s[2, 1]

        // second row
        invShearingMatrix[1, 0] = -s[0, 1] + s[2, 1] * s[0, 2]
        invShearingMatrix[1, 2] = -s[2, 1] + s[2, 0] * s[0, 1]

        // third row
        invShearingMatrix[2, 0] = -s[0, 2] + s[0, 1] * s[1, 2]
        invShearingMatrix[2, 1] = -s[1, 2] + s[1, 0] * s[0, 2]

        // divide by discriminant
        invShearingMatrix = invShearingMatrix.div(d)

        invMatrix *= invShearingMatrix
        forwardMatrix = s * forwardMatrix
    }

    override fun ray(ray: Ray): Ray {
        val origin = invMatrix * ray.origin
        val direction = invMatrix * ray.direction
        return Ray(origin, direction)
    }

    override fun equals(other: Any?): Boolean = this.equals<AffineTransformation>(other) { a, b ->
        a.forwardMatrix == b.forwardMatrix && a.invMatrix == b.invMatrix
    }

    override fun hashCode(): Int = this.hash(forwardMatrix, invMatrix)

    override fun toString(): String = "AffineTransformation($forwardMatrix, $invMatrix)"
}
